src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -330,18 +342,43 @@ private static int VerifyClientCertificate(int preverify_ok, IntPtr x509_ctx_ptr
             return OpenSslSuccess;
         }
 
-        private static unsafe int AlpnServerSelectCallback(IntPtr ssl, out IntPtr outp, out byte outlen, IntPtr inp, uint inlen, IntPtr arg)
+        private static unsafe int AlpnServerSelectCallback(IntPtr ssl, out byte* outp, out byte outlen, byte* inp, uint inlen, IntPtr arg)
         {
-            outp = IntPtr.Zero;
+            outp = null;
             outlen = 0;
-                        
-            GCHandle protocols = GCHandle.FromIntPtr(arg);
-            Debug.Assert(protocols.IsAllocated && protocols.Target != null);
 
-            byte[] server = (byte[])protocols.Target;
+            GCHandle protocolHandle = GCHandle.FromIntPtr(arg);
+            if (!(protocolHandle.Target is List<SslApplicationProtocol> protocolList))
+            {
+                return Ssl.SSL_TLSEXT_ERR_NOACK;
+            }
+
+            try
+            {
+                for (int i = 0; i < protocolList.Count; i++)
+                {
+                    Span<byte> clientList = new Span<byte>(inp, (int)inlen);
+                    while (clientList.Length > 0)
+                    {
+                        byte length = clientList[0];
+                        Span<byte> clientProto = clientList.Slice(1, length);
+                        if (clientProto.SequenceEqual(protocolList[i].Protocol.Span))
+                        {
+                            outp = inp + (int)inlen - clientList.Length + 1;
```

Comment:
I think that this is `outp = clientProto` (in pseudo-code).  Can you find a way of expressing that without math?

`outp = &clientProto.DangerousGetPinnableReference();`, maybe?

Reply:
Good change, no one likes pointer math if they can help it 👍

Reply:
But pity it won't work unless we wrap it in a fixed statement or use unsafe, which seems more messy

Reply:
FYI if you want a change where I take the dangerous ref and fix it to get the pointer... I can I just think it seems even more odd than the pointer math... but your call @bartonjs

Reply:
Is there anything on the Unsafe class which will help here?

Reply:
Could do
``` c#
outp = Unsafe.AsPointer(clientList.DangerousGetPinnableReference());
```
I will have to add `System.Runtime.CompilerServices` to the project I suspect, but if you are okay with that I do prefer that code to the math.